experimental ILD models for FCCee

danieljeans opened the pull request:
BEGINRELEASENOTES
- Fix severe overlaps in ILD_l5_v11 inner tracker (outer barrel layer)
- Introduce new experimental models ILD_FCCee_v01 and ILD_FCCee_v02 under FCCee/ILD_FCCee directory
-- MDI from FCCee common MDI area
-- vertex & lumi cal identical to CLD_o2_v07
-- for ILD_FCCee_v01: inner tracker now based on TrackerBarrel_o1_v06 (adapted from CLD_o2_v07), TPC inner radius slightly increased 329 -> 365 mm to avoid MDI envelope
-- for ILD_FCCee_v02: inner tracker identical to CLD_o2_v07, TPC inner radius significantly increased 329 -> 701 mm to accommodate larger inner tracker
-- no ECAL ring, LHCAL, beamcal
-- other subdetectors same as ILD_l5_v02
-- some moving/renaming of definitions in ILD_common_v02 was necessary
ENDRELEASENOTES

danieljeans commented:
@Victor-Schwan was heavily involved

Zehvogel commented:
You probably want to pick up also this LumiCal change: https://github.com/key4hep/k4geo/pull/391

Have you considered putting the ILD FCCee model into the FCCee directory? This might make its existence a bit more obvious :)

danieljeans commented:
@Zehvogel I had the same thought, we'll discuss it in ILD; maybe at least link it to the FCC directory.

tmadlener commented:
Looks like merging #388 introduced a conflict here (as far as I can see, it should be straight forward to fix by dropping the first commit). 

Did you also start to check whether this can be put into the `FCCee/ILD_FCCee` (or similar) either fully living there or via some clever symlinking?

danieljeans commented:
- fixed the conflicts (some missing definition...)
- moved the ILD_FCCee_v01 model to the FCCee directory. Sym-linking is non-trivial because main description files live at different depths under FCCee/ and ILD/ , which seems to create problems with relative paths...

tmadlener commented:
> Sym-linking is non-trivial because main description files live at different depths under FCCee/ and ILD/ , which seems to create problems with relative paths...

Yeah, this is what I was expecting. I think symlinking the `ILD_common_v02` to `FCCee/ILD_FCCee` is probably the best we can do, unless we want to explicitly copy the contents over.

danieljeans commented:
still to do: fix materials definitions in
`ILD/compact/ILD_common_v02/materials.xml `

which give rise to warnings:

```
Warning in <TGeoMixture::ComputeDerivedQuantities>: Mixture RPCGAS2: sum of weights is: 0.99249
Warning in <TGeoMixture::ComputeDerivedQuantities>: Mixture PEEK-GF30: sum of weights is: 1.03333
Warning in <TGeoMixture::ComputeDerivedQuantities>: Mixture g10-RPC: sum of weights is: 1.10025
```
have contacted expert (Gerald Grenier)

tmadlener commented:
Since nothing in this PR changes the materials definitions, those are pre-existing, right? In that case, I would prefer to fix that in a separate PR.

danieljeans commented:
> Since nothing in this PR changes the materials definitions, those are pre-existing, right? In that case, I would prefer to fix that in a separate PR.

OK, seems a good plan

tmadlener commented:
Minor request from my side: Could the contents of the `README.md`s that are currently in the `v01` and `v02` folders be put into one top level `README.md` in the `compact` top level folder? I think that would make it easier to see what the different models represent on a quick glance.

tmadlener reviewed (commented):
I am not sure I fully understand why the `vis` attributes need to be renamed, resp. where the collision is coming from. Shouldn't that just be an issue in case a `vis` property that is used by any detector is not defined? What am I missing here?

Victor-Schwan commented:
> I am not sure I fully understand why the `vis` attributes need to be renamed, resp. where the collision is coming from. Shouldn't that just be an issue in case a `vis` property that is used by any detector is not defined? What am I missing here?

We linked subdetectors from CLD. In the XML file of the CLD Inner Tracker `RedVis` is defined (see https://github.com/key4hep/k4geo/blob/main/FCCee%2FCLD%2Fcompact%2FCLD_o2_v07%2FInnerTracker_o2_v08.xml#L70) which causes a conflict with the definition in `ILD_common_v02`. Hence, we decided to rename ILDs `vis` attributes.

danieljeans commented on ILD/compact/ILD_common_v02/display.xml:
```diff
@@ -1,41 +1,37 @@
+    <vis name="ILD_VXDLayerVis"   alpha="1.0" r="0.1" g=".5"  b=".5"   showDaughters="true"  visible="true"/>
+    <vis name="ILD_VXDSupportVis" alpha="1.0" r="0.0" g="1.0" b="0.0"  showDaughters="true"  visible="true"/>
+
+    <vis name="ILD_FTDVis2"         alpha="1.0" r="0.0" g="0.1"  b="0.0"   showDaughters="true"  visible="false"/>
```
there were originally already definitions of FTDVis and ILD_FTDVis, so in this case (similarly for TPCVis) just adding ILD_ didn't work.

danieljeans commented on ILD/compact/README.md:
```diff
@@ -40,8 +41,16 @@ The following ILD detector models are available in lcgeo ( current production mo
 ### ILD_l5_v11
-same detector geometry as ILD_l5_v02 model, but with FCCee MDI and CLD-inspired silicon inner tracker & VTX in place of original ILD's VTX, SIT, FTD. 
+TEST model: same detector geometry as ILD_l5_v02 model, but with FCCee MDI and CLD-inspired silicon inner tracker & VTX in place of original ILD's VTX, SIT, FTD. 
 BeamCal, LHCAL removed.
```
Yes, I would recommend that. I added a note along those lines.

danieljeans commented on ILD/compact/ILD_common_v02/materials.xml:
```diff
@@ -434,6 +434,7 @@
       <MEE unit="eV" value="790"/>
       <D unit="g/cm3" value="19.32"/>
       <atom unit="g/mole" value="196.967"/>
+      <fraction n="1" ref="Au"/>
```
this was the easy part to fix. Sorry, it should probably best be combined with the other fixes, I'll revert it for now.

danieljeans commented on FCCee/ILD_FCCee/compact/ILD_FCCee_v01/FCCdefs_common.xml:
```diff
@@ -0,0 +1,103 @@
+
+
+  <!-- END info copied from CLD_o2_v07.xml -->  
+  <constant name="GlobalTrackerReadoutID" type="string" value="system:5,side:-2,layer:9,module:8,sensor:8"/>
```
hmm, not sure about this, to be checked.

danieljeans replied:
for other ILD subdetectors, it is hardcoded into the separate xml files for each system.
For tracking subdetectors it is always `system:5,side:2,layer:9,module:8,sensor:8` .

this "GlobalTrackerReadoutID" variable is used for the tracker elements taken from CLD (ie inner tracker, vertex detector). It seems that I adjusted it to (almost) match the ILD standard, but with `side:-2` (I think this was maybe the previous Mokka-time definition...I'm not sure why I did this...)
`system:5,side:-2,layer:6,module:11,sensor:8`
I don't see any reason not to change -2 -> 2 here.

Should we worry that 8 bits may not be enough for "module"?

On the other hand, I believe that @jfklama got track digi, conformal track reconstruction working with the v09 model, which used a mix of encodings, with for the outer tracker
`<id>system:5,side:-2,layer:6,module:11,sensor:8</id>`
Does conformal tracking use the cellID?

[TrackingDigi_SiILD.xml](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/Tracking/TrackingDigi_SiILD.xml)
[ConformalTracking_SiILD.xml](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/Tracking/ConformalTracking_SiILD.xml)
[HighLevelReco_SiILD.xml](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/HighLevelReco/HighLevelReco_SiILD.xml)

I'll look further next week.

danieljeans replied:
adjusted to 
` <constant name="GlobalTrackerReadoutID" type="string" value="system:5,side:2,layer:9,module:8,sensor:8"/>`
same as other ILD trackers

tmadlener replied:
I think the `8` module bits should be fine for us since there are less modules in the TPC region. I think that is the region where the number of modules will get larger for silicon detectors, such that they require 11 bits.

It's possible that conformal tracking can deal with different cell ID encodings better than the `FullLDCTracking` that is used for ILD.

danieljeans replied:
@tmadlener , @Victor-Schwan has got the conformal tracking working (at least to 1st order) with the silicon hits, and Clupatra on the TPC hits; linking the two collections is still to be done. So I think there is no fundamental problems caused by our cellID encodings.

tmadlener replied:
Nice :)

danieljeans commented on ILD/compact/ILD_common_v02/display.xml:
```diff
@@ -1,41 +1,37 @@
+    <vis name="ILD_BlackVis"         alpha="1.0" r="0.0" g="0.0"  b="0.0"   showDaughters="true"  visible="true"/>
+    <vis name="ILD_GrayVis"          alpha="1.0" r="0.5" g="0.5"  b="0.5"   showDaughters="true"  visible="true"/>
+
+    <vis name="ILD_TPCVis2"        alpha="0.1" r="0" g="0"  b="0"         showDaughters="true"  visible="true"/>
```
same reason as above

tmadlener commented:
> We linked subdetectors from CLD. In the XML file of the CLD Inner Tracker `RedVis` is defined (see https://github.com/key4hep/k4geo/blob/main/FCCee%2FCLD%2Fcompact%2FCLD_o2_v07%2FInnerTracker_o2_v08.xml#L70) which causes a conflict with the definition in `ILD_common_v02`. Hence, we decided to rename ILDs vis attributes.

I see, that makes sense. Thanks. I suppose there is no easy way of leaving the other models untouched, e.g. by copying the `display.xml` to an `ILD_FCCee_common` folder, and removing the clashes? Would that not work because then the ILD parts would be missing some of the vis attibutes?

tmadlener commented on FCCee/ILD_FCCee/compact/ILD_FCCee_v01/basic_defs.xml:
```diff
@@ -0,0 +1,44 @@
+  <constant name="ILDDetID_NOTUSED"      value="  0"/>
+
+
+  <constant name="ILDDetID_TPC"          value="  4"/>
```
```suggestion
  <constant name="ILDDetID_TPC"          value="  4"/>
```
I think we should maybe put in a comment here as well, that states that `4` is what is expected for the TPC (at least for now)

danieljeans replied:
does this stem from the definition in 
[lcio/UTIL/ILDConf](https://github.com/iLCSoft/LCIO/blob/master/src/cpp/src/UTIL/ILDConf.cc) ?

danieljeans replied:
added comment

tmadlener replied:
Yes, I think @Victor-Schwan and me figured out that that is the source of that requirement.

danieljeans commented:
> > 
> 
> I see, that makes sense. Thanks. I suppose there is no easy way of leaving the other models untouched, e.g. by copying the `display.xml` to an `ILD_FCCee_common` folder, and removing the clashes? Would that not work because then the ILD parts would be missing some of the vis attibutes?

There is certainly scope for sharing colour definitions among different k4geo models, but it would take some effort...

Victor-Schwan commented on FCCee/ILD_FCCee/compact/ILD_FCCee_v02/basic_defs.xml:
Well-spotted! How about moving the shared file up to the parent directory?

danieljeans replied:
https://github.com/key4hep/k4geo/pull/390/commits/25b5ddfddc3b49205cf1ea7943f316932d546ad1 avoids cross-linking between v01 and v02 models: shared definitions moved to dedicated directory.

tmadlener commented on ILD/compact/README.md:
```diff
@@ -40,9 +41,18 @@ The following ILD detector models are available in lcgeo ( current production mo
-same detector geometry as ILD_l5_v02 model, but with FCCee MDI and CLD-inspired silicon inner tracker & VTX in place of original ILD's VTX, SIT, FTD. 
+TEST DEVELOPMENT model: same detector geometry as ILD_l5_v02 model, but with FCCee MDI and CLD-inspired silicon inner tracker & VTX in place of original ILD's VTX, SIT, FTD. 
 BeamCal, LHCAL removed.
+We recommend that you use an ILD model under FCCee/ILD_FCCee/ instead. Model ILD_FCCee_v01 is rather similar to ILD_l5_v11.
```
```suggestion
**We recommend that you use an ILD model under FCCee/ILD_FCCee/ instead. Model ILD_FCCee_v01 is rather similar to ILD_l5_v11.**
```
Just to make it stick out a bit more.

tmadlener commented:
What is still missing to lift this out of the WIP status that it's currently in?

Victor-Schwan commented:
> What is still missing to lift this out of the WIP status that it's currently in?
#381 introduces changes to the MDI e.g. to the beampipe in the common_MDI v00 and we rebased on top of that PR. Hence, we thought to merge the other PR first. I am not sure whether the changes are needed though

danieljeans commented:
> > What is still missing to lift this out of the WIP status that it's currently in?
> 
> #381 introduces changes to the MDI e.g. to the beampipe in the common_MDI v00. We rebased on top of that PR and linked several files there. Hence, we thought to merge the other PR first. I am not sure whether the changes are needed though

@Victor-Schwan are you sure we are based on that PR? I think it works fine on the master branch (eg the lumical is taken from the CLD directory). If that MDI PR goes through, we would update to profit from improvements/changes.
In my understanding, we're ready to remove WIP here (on the understanding that they are experimental models and not an "officially blessed" version of ILD).

Victor-Schwan commented:
> > > What is still missing to lift this out of the WIP status that it's currently in?
> > 
> > 
> > #381 introduces changes to the MDI e.g. to the beampipe in the common_MDI v00. We rebased on top of that PR and linked several files there. Hence, we thought to merge the other PR first. I am not sure whether the changes are needed though
> 
> @Victor-Schwan are you sure we are based on that PR? I think it works fine on the master branch (eg the lumical is taken from the CLD directory). If that MDI PR goes through, we would update to profit from improvements/changes. In my understanding, we're ready to remove WIP here (on the understanding that they are experimental models and not an "officially blessed" version of ILD).

I checked again and initially, I rebased another branch with `v11` on the `common mdi` PR. I did not see any problem with this PR. Sorry for the confusion

tmadlener commented:
> ready to remove WIP here (on the understanding that they are experimental models and not an "officially blessed" version of ILD)

I think it's stated quite obviously in the provided README and to really make sure we could also put it into the release notes once more. For me this seems to be good as it is, also because there will likely be some iteration still once we have reconstruction available. So, as soon as you remove the *WIP* I would be fine to have a last look and merge.

danieljeans commented:
WIP removed

tmadlener reviewed (approved):
This is good to merge from my side. Paging @BrieucF and @andresailer to make sure that they are also fine with putting this under the FCC umbrella.

BrieucF commented:
> This is good to merge from my side. Paging @BrieucF and @andresailer to make sure that they are also fine with putting this under the FCC umbrella.

This is fine by me!
